relay-server/src/service.rs

```diff
@@ -320,30 +322,52 @@ where
     }
 }
 
-/// Given a relay config spawns the server together with all actors and lets them run forever.
-///
-/// Effectively this boots the server.
-pub fn start(config: Config) -> Result<Recipient<server::StopServer>> {
-    let config = Arc::new(config);
-
-    Controller::from_registry().do_send(Configure {
-        shutdown_timeout: config.shutdown_timeout(),
-    });
-
-    let state = ServiceState::start(config.clone())?;
-    let mut server = server::new(move || make_app(state.clone()));
-    server = server
-        .workers(config.cpu_concurrency())
-        .shutdown_timeout(config.shutdown_timeout().as_secs() as u16)
-        .keep_alive(config.keepalive_timeout().as_secs() as usize)
-        .maxconn(config.max_connections())
-        .maxconnrate(config.max_connection_rate())
-        .backlog(config.max_pending_connections())
-        .disable_signals();
-
-    server = listen(server, &config)?;
-    server = listen_ssl(server, &config)?;
-
-    dump_listen_infos(&server);
-    Ok(server.start().recipient())
+/// Keeps the address to the running http servers and helps with start/stop handling.
+#[derive(Clone)]
+pub struct HttpServer(Recipient<StopServer>);
+
+impl HttpServer {
+    /// Given a relay config spawns the server together with all actors and lets them run forever.
+    ///
+    /// Effectively this boots the server.
+    pub fn start(config: Config) -> Result<Self> {
+        let config = Arc::new(config);
+
+        Controller::from_registry().do_send(Configure {
+            shutdown_timeout: config.shutdown_timeout(),
+        });
+
+        let state = ServiceState::start(config.clone())?;
+        let mut server = server::new(move || make_app(state.clone()));
+        server = server
+            .workers(config.cpu_concurrency())
+            .shutdown_timeout(config.shutdown_timeout().as_secs() as u16)
+            .keep_alive(config.keepalive_timeout().as_secs() as usize)
+            .maxconn(config.max_connections())
+            .maxconnrate(config.max_connection_rate())
+            .backlog(config.max_pending_connections())
+            .disable_signals();
+
+        server = listen(server, &config)?;
+        server = listen_ssl(server, &config)?;
+
+        dump_listen_infos(&server);
+        let recipient = server.start().recipient();
+        Ok(Self(recipient))
+    }
+
+    /// Triggers the shutdown process by sending [`actix_web::server::StopServer`] to the running http server.
+    pub fn shutdown(self, graceful: bool) {
```

Comment:
Why does this take `self` by value instead of by reference?

Reply:
I guess it mostly to signal to the caller that once you can the shutdown, the server is not longer can be used after that.

Reply:
The intent makes sense, but wouldn't changing it to `&mut self` prevent the need to clone [here](https://github.com/getsentry/relay/pull/1723/files#diff-a37078486e7ec6da20e15d45dc8c8cec32ae64b648bdd086d0287274336098b0R28)?

Reply:
you're right. I've changed it.